### PR TITLE
Implement sass_layer() and sass_layers()

### PR DESCRIPTION
**Pull request by cpsievert**

### CODE REVIEW

File: R/as_sass.R [hunk 1]
```diff
@@ -105,6 +105,11 @@ as_sass_.list <- function(input) {
   paste0("$", input_names, ": ", input_values, ";", collapse = "\n")
 }
 
+as_sass_.sass_layer <- function(input) {
+  sass <- as_sass_(list(input$pre, input$post))
+  attachDependencies(as_html(sass, "sass"), input$deps)
```
Comment (jcheng5) [inline R/as_sass.R]: Dependencies don't need to be attached to the result, `sass::sass` is the consumer of these objects and it doesn't care

Comment (jcheng5) [inline R/as_sass.R]: Don't think `as_html` is needed here nor in `as_sass` (though they do need to be `class="sass"`)

Comment (cpsievert) [reply R/as_sass.R]: Should `sass::sass()` attempt to retrieve and process the dependencies? It feels a bit odd to me that we'd have `sass_layers()` understand and accumulate dependencies, but then ultimately do nothing useful with that knowledge in `sass::sass()`

File: R/layers.R [hunk 2]
```diff
@@ -72,3 +72,26 @@ is_sass_layer <- function(x) {
 dropNulls <- function(x) {
   x[!vapply(x, is.null, FUN.VALUE=logical(1))]
 }
+
+# currently sass_layers are the only thing that
+# have a notion of html dependencies
+html_dependencies <- function(x, deps = NULL) {
```
Comment (jcheng5) [inline R/layers.R]: Remove `deps` argument

File: R/layers.R [hunk 3]
```diff
@@ -72,3 +72,26 @@ is_sass_layer <- function(x) {
...
+    html_deps <- list(html_deps)
+  }
+  deps <- c(deps, html_deps)
+
+  if (is_sass_layer(x)) {
+    deps <- c(deps, x$deps)
+  }
+
+  # recursive case
+  if (is.list(x)) {
```
Comment (jcheng5) [inline R/layers.R]: `else if`?

Comment (jcheng5) [reply R/layers.R]: Or not.

File: R/sass.R [hunk 4]
```diff
@@ -96,11 +96,19 @@ sass <- function(input = NULL, options = sass_options(), output = NULL,
   }
 
   css <- as_html(css, "css")
+  deps <- html_dependencies(input)
+  deps <- htmltools::resolveDependencies(deps)
 
   if (!is.null(output)) {
     writeLines(css, output)
+    if (length(deps)) {
+      lapply(deps, htmltools::copyDependencyToDir, outputDir = dirname(output))
```
Comment (jcheng5) [inline R/sass.R]: Don't do this, it drops the identity of the dependencies you're copying to disk.

File: R/sass.R [hunk 5]
```diff
@@ -96,11 +96,19 @@ sass <- function(input = NULL, options = sass_options(), output = NULL,
...
+    }
     return(invisible())
   } else {
```
Comment (jcheng5) [inline R/sass.R]: Unconditionally attach dependencies and return CSS, just do it invisibly if `!is.null(output)`.

File: R/sass.R [hunk 6]
```diff
@@ -96,11 +96,19 @@ sass <- function(input = NULL, options = sass_options(), output = NULL,
   }
 
   css <- as_html(css, "css")
+  deps <- html_dependencies(input)
+  deps <- htmltools::resolveDependencies(deps)
```
Comment (jcheng5) [inline R/sass.R]: Either don't resolve dependencies, or, pass `FALSE` for `resolvePackageDir` argument

File: tests/testthat/test-html-dependencies.R [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+test_that("sass() relays sass_layer()'s html dependencies", {
+
+  input <- list(
+    sass_layer(pre = "body{color: green}", deps = dep1),
+    "body{color: red}"
+  )
+  expect_equal(htmlDependencies(sass(input)), list(dep1))
+
+  # if output is written to a file, the dependency should be copied
+  # over to that output directory
```
Comment (jcheng5) [inline tests/testthat/test-html-dependencies.R]: Remove

File: tests/testthat/test-html-dependencies.R [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  output2 <- sass(list(list(input1)))
+  output3 <- sass(list(list(input1), ".foo{color: red}", list(list(input1))))
+  expect_equal(htmlDependencies(output1), list(dep1))
+  expect_equal(htmlDependencies(output2), list(dep1))
+  expect_equal(htmlDependencies(output3), list(dep1))
+  # multiple html dependencies
+  input2 <- attachDependencies("body {color: red}", dep2)
+  output4 <- sass(list(list(input1), list(), list(list(input2))))
+  expect_equal(htmlDependencies(output4), list(dep1, dep2))
+})
```
Comment (jcheng5) [inline tests/testthat/test-html-dependencies.R]: Can you add a test that uses `sass_layer_merge` with dependencies?

**Pull request merged**